### PR TITLE
Implement Phidget Manager to make Phidget discovery a bit simpler

### PR DESCRIPTION
**Pull request by AndreRenaud**
This adds a new PhidgetManager implementation, and expands the Phidget implementation to get a bit more info about them to make printing easier

### CODE REVIEW

**Comment by jrcichra** [comment]
I'll give this a try on my network of phidgets and see what it discovers

**Comment by jrcichra** [comment]
I get `found 0 phidgets` pretty quickly in the example. I've never used the Manager before. Does it work for you?

**Comment by AndreRenaud** [comment]
Yes, it works for me. But I've only got a single Phidget which is directly attached.
If you put a long sleep before calling ListPhidgets, does it improve?
```
% go run main.go
Found 9 phidgets
  0: Thermocouple Input: PhidgetTemperatureSensor channel 0 [ser=8b85c] [label=] [port=0]
Temperature is 28.230600
  1: Thermocouple Input: PhidgetTemperatureSensor channel 1 [ser=8b85c] [label=] [port=0]
Temperature is 25.369600
  2: Thermocouple Input: PhidgetTemperatureSensor channel 2 [ser=8b85c] [label=] [port=0]
Temperature is 402.675100
  3: Thermocouple Input: PhidgetTemperatureSensor channel 3 [ser=8b85c] [label=] [port=0]
Temperature is 403.169800
  4: Temperature Sensor (IC): PhidgetTemperatureSensor channel 4 [ser=8b85c] [label=] [port=0]
Temperature is 25.789100
  5: Voltage Input: PhidgetVoltageInput channel 0 [ser=8b85c] [label=] [port=0]
Voltage is 0.000099
  6: Voltage Input: PhidgetVoltageInput channel 1 [ser=8b85c] [label=] [port=0]
Voltage is -0.000019
  7: Voltage Input: PhidgetVoltageInput channel 2 [ser=8b85c] [label=] [port=0]
Voltage is 0.015478
  8: Voltage Input: PhidgetVoltageInput channel 3 [ser=8b85c] [label=] [port=0]
Voltage is 0.015499
```

**Comment by jrcichra** [comment]
I might have an older copy of phidgets22.h on my desktop. Let me recompile the latest. It's working on a raspberry pi I have that has the USB HUB1000? plugged into it:

```
unsupported phidget discovered: 0xd
unsupported phidget discovered: 0xf
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x5
unsupported phidget discovered: 0x6
unsupported phidget discovered: 0x6
unsupported phidget discovered: 0x6
unsupported phidget discovered: 0x6
unsupported phidget discovered: 0x6
unsupported phidget discovered: 0x6
Found 18 phidgets
  0: Current Input: PhidgetCurrentInput channel 0 [ser=8962a] [label=] [port=0]
  1: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=1]
Voltage is 12.773900
  2: Current Input: PhidgetCurrentInput channel 0 [ser=8962a] [label=] [port=2]
  3: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=3]
Voltage is 0.107700
  4: Temperature Sensor: PhidgetTemperatureSensor channel 0 [ser=8962a] [label=] [port=4]
Temperature is 12.630000
  5: Light Sensor: PhidgetLightSensor channel 0 [ser=8962a] [label=] [port=5]
  6: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=0]
Voltage is 2.864230
  7: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=1]
Voltage is 2.708890
  8: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=2]
Voltage is 2.866970
  9: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=3]
Voltage is 2.707060
  10: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=4]
Voltage is 2.178190
  11: Voltage Input: PhidgetVoltageInput channel 0 [ser=8962a] [label=] [port=5]
unsupported phidget discovered: 0xf
Voltage is 2.868650
  12: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=0]
  13: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=1]
  14: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=2]
  15: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=3]
  16: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=4]
  17: Voltage Ratio Input: PhidgetVoltageRatioInput channel 0 [ser=8962a] [label=] [port=5]

```

**Comment by jrcichra** [comment]
All my experiments have been with a `time.Sleep(20)` between new and list. I can't get it to work on devices that don't have Phidgets physically plugged into them. But the one that did have Phidgets physically plugged in, found 18 Phidgets. I don't believe the Go wrapper is to blame. I'm good with this change.

**Comment by AndreRenaud** [comment]
The PhidgetManager currently has a 20ms sleep to try and get things discovered before it returns. I found this purely by guesswork, so we could try bumping it to 200ms or something. I think anything much larger than that would be annoying for users (noticeably slow).

File: phidgets/manager.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	// TODO: Should install a PhidgetManager_OnDetachCallback as well so we can
+	// support removing the devices
+
+	if cerr := C.PhidgetManager_open(m.handle); cerr != C.EPHIDGET_OK {
+		C.PhidgetManager_delete(&m.handle)
+		return nil, managerError(cerr)
+	}
+
+	// Sometimes the phidgets take a while to attach
+	time.Sleep(20 * time.Millisecond)
```
Comment (jrcichra) [inline phidgets/manager.go:110]: Maybe instead, this sleep should be left for the user to implement. Going off the example here: https://www.phidgets.com/docs/Phidget_Manager they'd have to sleep if they want to poll with `ListPhidgets` immediately after `open`.

Not having the sleep shouldn't affect the attach handler, right?

Comment (AndreRenaud) [reply phidgets/manager.go:110]: Yes, that wouldn't be unreasonable. It's annoying that existing Phidgets aren't immediately available, but this is really a feature of the Phidget API, not the Go library, so we should just replicate their functionality. 20ms sleep isn't accurate anyway. I'll remove this and put an explicity sleep in main.go instead, with a comment about it.

File: phidgets/phidget.go [hunk 2]
```diff
@@ -129,3 +142,44 @@ func (p *phidget) GetChannelClassName() (string, error) {
...
+	return C.GoString(cstr), nil
+}
+
+// GetChannelName retrieves the channel name of this handle
+func (p *phidget) GetChannelName() (string, error) {
+	var cstr *C.char
+	if cerr := C.Phidget_getChannelName(p.handle, &cstr); cerr != C.EPHIDGET_OK {
+		return "", p.phidgetError(cerr)
+	}
+	return C.GoString(cstr), nil
```
Comment (jrcichra) [inline phidgets/phidget.go:161]: There's no memory leak here right? `cstr` is just a Go pointer on the stack where `C.Phdget_getChannelName` gives us an address? So Go frees the additional address pointer (`cstr`) automatically.

Comment (AndreRenaud) [reply phidgets/phidget.go:161]: No, `C.GoString` does not take ownership of the underlying memory, I think it duplicate it, so it is up to the caller to make sure things are free'd properly. However `C.Phidget_getChannelName` does not returned a malloc'd pointer, it is just a static string, so there is nothing to free.

**Comment by jrcichra** [comment]
I'm satisfied to squash/merge. Any other concerns we should address?. The `TODO`s mentioned can come in future pull requests.

**Comment by AndreRenaud** [comment]
I'm ok with it now - agreed on the TODOs. I've got an order of Phidgets arriving early next week, so I'll probably add a couple of driver modules in then.

**Pull request merged**